File: apps/gas/src/handlers/planner_monthly.ts
/**
 * スピードプランナー「月間管理」読取ハンドラ
 * - 読み取り専用: 指定 (year, month) の行をフィルタして返す
 */
import { ApiResponse, ok, ng, toNumberOrNull } from "../lib/common";
import { resolveSpreadsheetIdByStudent } from "../lib/student_resolver";
import { MONTHLY_SHEET_NAME } from "../lib/columns";

type RowMap = Record<string, any>;

// openMonthlySheet uses shared resolveSpreadsheetIdByStudent from lib/student_resolver.ts
function openMonthlySheet(req: RowMap): GoogleAppsScript.Spreadsheet.Sheet | null {
  const fid = req.spreadsheet_id || resolveSpreadsheetIdByStudent(req);
  if (!fid) return null;
  const ss = SpreadsheetApp.openById(fid);
  const sh = ss.getSheetByName(MONTHLY_SHEET_NAME);
  return sh || null;
}

function normYear2(y: any): number | null {
  const n = Number(String(y).trim());
  if (!Number.isFinite(n)) return null;
  if (n >= 2000) return n - 2000; // 2025→25
  if (n >= 0 && n <= 99) return n; // 0..99はそのまま
  return null;
}

/**
 * planner.monthly.filter
 * 入力: student_id? or spreadsheet_id?, year(2桁/4桁), month(1..12)
 */
export function plannerMonthlyFilter(req: RowMap): ApiResponse {
  const { year, month } = req;
  const yy = normYear2(year);
  const mm = Number(String(month || "").trim());
  if (yy == null || !Number.isFinite(mm) || mm < 1 || mm > 12) {
    return ng("planner.monthly.filter", "BAD_REQUEST", "year(2桁/4桁) と month(1..12) を指定してください");
  }
  const sh = openMonthlySheet(req);
  if (!sh) return ng("planner.monthly.filter", "NOT_FOUND", "monthly sheet not found (月間管理)");

  const lastRow = sh.getLastRow();
  if (lastRow <= 1) return ok("planner.monthly.filter", { year: yy, month: mm, items: [], count: 0 });

  // 2行目〜最終行の A..R を display で取得
  const numRows = lastRow - 1;
  const values = sh.getRange(2, 1, numRows, 18).getDisplayValues();

  const items: any[] = [];
  for (let i = 0; i < values.length; i++) {
    const r = i + 2; // シート上の行番号
    const row = values[i];
    const A = String(row[0] || "");
    const B = String(row[1] || "").trim();
    const C = String(row[2] || "").trim();
    if (!A && !B && !C) continue; // 完全空行
    const bNum = Number(B);
    const cNum = Number(C);
    if (bNum !== yy || cNum !== mm) continue;
    const K = toNumberOrNull(row[10]);
    const L = toNumberOrNull(row[11]);
    const M = toNumberOrNull(row[12]);
    const weeks = [13,14,15,16,17].map((colIdx, j) => ({ index: j+1, actual: String(row[colIdx] || "") }));
    const monthCode = (Number.isFinite(bNum) && Number.isFinite(cNum)) ? (bNum*10 + cNum) : null;
    items.push({
      row: r,
      raw_code: A,
      month_code: monthCode,
      year: bNum,
      month: cNum,
      book_id: String(row[6] || ""), // G
      subject: String(row[7] || ""), // H
      title: String(row[8] || ""),   // I
      guideline_note: String(row[9] || ""), // J
      unit_load: K,
      monthly_minutes: L,
      guideline_amount: M,
      weeks,
    });
  }
  return ok("planner.monthly.filter", { year: yy, month: mm, items, count: items.length });
}

File: apps/mcp/handlers/planner_monthly.py
"""
Planner (monthly) handler.
Ported from apps/gas/src/handlers/planner_monthly.ts

Provides reading operations for monthly planner sheets.
"""
from typing import Any

from sheets_client import SheetsClient
from config import MONTHLY_SHEET_NAME
from lib.common import ok, ng, to_number_or_none
from .planner import resolve_planner_sheet_id


def _norm_year_2digit(year: Any) -> int | None:
    """Normalize year to 2-digit format (e.g., 2025 -> 25)."""
    try:
        n = int(str(year).strip())
        if n >= 2000:
            return n - 2000
        if 0 <= n <= 99:
            return n
        return None
    except (ValueError, TypeError):
        return None


def planner_monthly_filter(
    sheets: SheetsClient,
    year: int,
    month: int,
    student_id: str | None = None,
    spreadsheet_id: str | None = None,
) -> dict:
    """
    Filter monthly planner data by year and month.

    Args:
        year: Year (2-digit or 4-digit)
        month: Month (1-12)
        student_id: Student ID for resolving spreadsheet
        spreadsheet_id: Direct spreadsheet ID

    Returns:
        Items matching the specified year/month
    """
    yy = _norm_year_2digit(year)
    try:
        mm = int(str(month).strip())
        if mm < 1 or mm > 12:
            mm = None
    except (ValueError, TypeError):
        mm = None

    if yy is None or mm is None:
        return ng("planner.monthly.filter", "BAD_REQUEST", "year(2桁/4桁) と month(1..12) を指定してください")

    resolved_id = resolve_planner_sheet_id(sheets, student_id, spreadsheet_id)
    if not resolved_id:
        return ng("planner.monthly.filter", "NOT_FOUND", "monthly sheet not found (月間管理)")

    try:
        ss = sheets.open_by_id(resolved_id)
        ws = ss.worksheet(MONTHLY_SHEET_NAME)
    except Exception as e:
        return ng("planner.monthly.filter", "NOT_FOUND", f"monthly sheet not found: {e}")

    try:
        # Read from row 2 to the end, columns A to R (18 columns)
        all_values = ws.get_all_values()
        if len(all_values) <= 1:
            return ok("planner.monthly.filter", {"year": yy, "month": mm, "items": [], "count": 0})

        values = all_values[1:]  # Skip header row

        items = []
        for i, row in enumerate(values):
            r = i + 2  # Sheet row number

            # Extract basic columns
            a = str(row[0]) if len(row) > 0 else ""
            b = str(row[1]).strip() if len(row) > 1 else ""
            c = str(row[2]).strip() if len(row) > 2 else ""

            if not a and not b and not c:
                continue  # Skip empty rows

            # Parse year/month from B/C columns
            try:
                b_num = int(b) if b else None
                c_num = int(c) if c else None
            except ValueError:
                continue

            if b_num != yy or c_num != mm:
                continue

            # Extract additional columns (G-R)
            book_id = str(row[6]) if len(row) > 6 else ""      # G
            subject = str(row[7]) if len(row) > 7 else ""      # H
            title = str(row[8]) if len(row) > 8 else ""        # I
            guideline_note = str(row[9]) if len(row) > 9 else ""  # J
            unit_load = to_number_or_none(row[10]) if len(row) > 10 else None  # K
            monthly_minutes = to_number_or_none(row[11]) if len(row) > 11 else None  # L
            guideline_amount = to_number_or_none(row[12]) if len(row) > 12 else None  # M

            # Week columns (N-R)
            weeks = []
            for j, col_idx in enumerate([13, 14, 15, 16, 17]):
                actual = str(row[col_idx]) if len(row) > col_idx else ""
                weeks.append({"index": j + 1, "actual": actual})

            # Calculate month code
            month_code = (b_num * 10 + c_num) if b_num is not None and c_num is not None else None

            items.append({
                "row": r,
                "raw_code": a,
                "month_code": month_code,
                "year": b_num,
                "month": c_num,
                "book_id": book_id,
                "subject": subject,
                "title": title,
                "guideline_note": guideline_note,
                "unit_load": unit_load,
                "monthly_minutes": monthly_minutes,
                "guideline_amount": guideline_amount,
                "weeks": weeks,
            })

        return ok("planner.monthly.filter", {
            "year": yy,
            "month": mm,
            "items": items,
            "count": len(items),
        })
    except Exception as e:
        return ng("planner.monthly.filter", "ERROR", str(e))
